nexus/src/app/sagas/region_snapshot_replacement_step.rs

```diff
@@ -0,0 +1,603 @@
+//! saga's responsibility to update all Volumes that reference the region
+//! snapshot being replaced, and send a replacement request to any Upstairs that
+//! were constructed.
+//!
+//! Some difficulty comes from the requirement to notify existing Upstairs that
+//! reference the replaced read-only part, but even this is not as difficult as
+//! region replacement: Nexus does not have to continually monitor and drive
+//! either live repair or reconciliation, just ensure that the read-only
+//! replacement occurs. Read-only replacements should be basically
+//! instantaneous.
+//!
+//! A replace request only needs to be done once per Upstairs that has the old
+//! reference. This is done as a "region snapshot replacement step", and once
+//! all those are done, the region snapshot replacement request can be
+//! "completed".
+//!
+//! Region snapshot replacement steps need to be written into the database and
+//! have an associated state and operating saga id for the same reason that
+//! region snapshot replacement requests do: multiple background tasks will
+//! invoke multiple sagas, and there needs to be some exclusive access.
+//!
+//! See the documentation for the "region snapshot replacement step garbage
+//! collect" saga for the next step in the process.
+
+use super::{
+    ActionRegistry, NexusActionContext, NexusSaga, SagaInitError,
+    ACTION_GENERATE_ID,
+};
+use crate::app::db::datastore::ExistingTarget;
+use crate::app::db::datastore::ReplacementTarget;
+use crate::app::db::datastore::VolumeToDelete;
+use crate::app::db::datastore::VolumeWithTarget;
+use crate::app::db::lookup::LookupPath;
+use crate::app::sagas::declare_saga_actions;
+use crate::app::{authn, authz, db};
+use nexus_db_model::VmmState;
+use nexus_types::identity::Resource;
+use omicron_common::api::external::Error;
+use propolis_client::types::ReplaceResult;
+use serde::Deserialize;
+use serde::Serialize;
+use sled_agent_client::types::CrucibleOpts;
+use sled_agent_client::types::VolumeConstructionRequest;
+use std::net::SocketAddrV6;
+use steno::ActionError;
+use steno::Node;
+use uuid::Uuid;
+
+// region snapshot replacement step saga: input parameters
+
+#[derive(Debug, Deserialize, Serialize)]
+pub(crate) struct Params {
+    pub serialized_authn: authn::saga::Serialized,
+    pub request: db::model::RegionSnapshotReplacementStep,
+}
+
+// region snapshot replacement step saga: actions
+
+declare_saga_actions! {
+    region_snapshot_replacement_step;
+    SET_SAGA_ID -> "unused_1" {
+        + rsrss_set_saga_id
+        - rsrss_set_saga_id_undo
+    }
+    CREATE_REPLACE_PARAMS -> "replace_params" {
+        + rsrss_create_replace_params
+    }
+    CREATE_FAKE_VOLUME -> "unused_2" {
+        + rssrs_create_fake_volume
+        - rssrs_create_fake_volume_undo
+    }
+    REPLACE_SNAPSHOT_IN_VOLUME -> "unused_3" {
+        + rsrss_replace_snapshot_in_volume
+        - rsrss_replace_snapshot_in_volume_undo
```

Comment:
Yep, it's possible that a user could issue a snapshot request to one of these disks after the volume replacement occurs, but if that occurs then the read-only parent of the snapshot they just created will reference the replacement, not the one being replaced.

If the snapshot is taken before the volume replacement occurs, then a region snapshot replacement step record will be created in the future for that resulting new volume because it would be referencing the one being replaced.